src/audio/igo_nr/igo_nr.c

```diff
@@ -440,67 +401,55 @@ static int32_t igo_nr_params(struct comp_dev *dev,
 
 	return cd->invalid_param ? -EINVAL : 0;
 }
+#endif
 
-static inline void igo_nr_set_chan_process(struct comp_dev *dev, int32_t chan)
+static inline void igo_nr_set_chan_process(struct comp_data *cd, int32_t chan)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-
 	if (!cd->process_enable[chan])
 		cd->process_enable[chan] = true;
 }
 
-static inline void igo_nr_set_chan_passthrough(struct comp_dev *dev, int32_t chan)
+static inline void igo_nr_set_chan_passthrough(struct comp_data *cd, int32_t chan)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-
 	if (cd->process_enable[chan]) {
 		cd->process_enable[chan] = false;
 		IgoLibInit(cd->p_handle, &cd->igo_lib_config, &cd->config.igo_params);
 	}
 }
 
-static int32_t igo_nr_cmd_get_data(struct comp_dev *dev,
-				   struct sof_ipc_ctrl_data *cdata,
-				   int32_t max_size)
+static int igo_nr_get_config(struct processing_module *mod,
+			     uint32_t config_id, uint32_t *data_offset_size,
+			     uint8_t *fragment, size_t fragment_size)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
-	int32_t ret;
+	struct comp_dev *dev = mod->dev;
 
-	switch (cdata->cmd) {
-	case SOF_CTRL_CMD_BINARY:
-		comp_info(dev, "igo_nr_cmd_get_data(), SOF_CTRL_CMD_BINARY");
-		ret = comp_data_blob_get_cmd(cd->model_handler, cdata, max_size);
-		break;
-	default:
-		comp_err(dev, "igo_nr_cmd_get_data() error: invalid cdata->cmd %u", cdata->cmd);
-		ret = -EINVAL;
-		break;
-	}
-	return ret;
-}
-
-static int32_t igo_nr_cmd_get_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata)
-{
-	struct comp_data *cd = comp_get_drvdata(dev);
-	int32_t j;
-	int32_t ret = 0;
+#if CONFIG_IPC_MAJOR_3
+	struct comp_data *cd = module_get_private_data(mod);
+	struct sof_ipc_ctrl_data *cdata = (struct sof_ipc_ctrl_data *)fragment;
+	int j;
 
 	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "igo_nr_get_config(), SOF_CTRL_CMD_BINARY");
+		return comp_data_blob_get_cmd(cd->model_handler, cdata, fragment_size);
 	case SOF_CTRL_CMD_SWITCH:
 		for (j = 0; j < cdata->num_elems; j++) {
 			cdata->chanv[j].channel = j;
 			cdata->chanv[j].value = cd->process_enable[j];
-			comp_info(dev, "igo_nr_cmd_get_value(), channel = %u, value = %u",
+			comp_info(dev, "igo_nr_get_config(), channel = %u, value = %u",
 				  cdata->chanv[j].channel,
 				  cdata->chanv[j].value);
 		}
-		break;
+		return 0;
 	default:
-		comp_err(dev, "igo_nr_cmd_get_value() error: invalid cdata->cmd %d", cdata->cmd);
-		ret = -EINVAL;
-		break;
+		comp_err(dev, "igo_nr_cmd_get_config() error: invalid cdata->cmd %d", cdata->cmd);
+		return -EINVAL;
 	}
-	return ret;
+
+#elif CONFIG_IPC_MAJOR_4
+	comp_err(dev, "igo_nr_cmd_get_config() error: not supported");
```

Comment:
@ranj063 How should we handle get_config() with IPC4, error it like here, or add code to populate the struct with response as in IPC3? If a control is volatile a get_config() to firmware might happen. I tried it today, but kernel doesn't allow volatile controls for dynamic pipelines. So, is it best to just return error if this would get called?

Reply:
If not supported **today** then return an error, we can always come back when kernel/topology ready.

Reply:
Yep, we discussed this with @ujfalusi , the control get to firmware is not happening unless we update the kernel to handle volatile type. The only case is when FW sends an empty control notification to which kernel responds with control get. But IGO NR code is not sending notifications.

Reply:
Thats fine, please just proceed with what we have and we can update for IPC4 when kernel ready.